src/nvim/ex_cmds.lua

```diff
@@ -2933,6 +2933,12 @@ module.cmds = {
     addr_type='ADDR_OTHER',
     func='ex_tag',
   },
+  {
+    command='trust',
```

Comment:
- AFAIK Vim's decided convention for command meta-args is the `++` prefix. Thus `:trust ++allow ...`, `:trust ++deny`. 
    - I'm basing this on discussion from when `:autocmd ... ++nested` was added.
    - Other than the old `:cscope` command, are there other prominent cases of the "subcommand" pattern in vim ex commands? If not, we should probably stick to the `++arg` convention.
- Could we get away with aliasing `:trust foo` to mean `:trust ++allow foo` or is that too clever?

Reply:
@ii14 That is essentially what the `view` option of `vim.secure.read()` already does, minus the `:Trust` user command. If possible I'd like to avoid using a user command (i.e. upper case command) in "first party" code as that claims that command from all plugins that may want to use it.

That also still doesn't answer the question of how to provide a "deny" and "delete"/"forget" function to `:trust`.

Reply:
Sure, but the command is my entire point :) The user command is just for demonstration purposes, the lowercase `:trust` could do that to make sure that you know exactly what you're whitelisting.

Reply:
> "subcommand" pattern

There are `:sign`, `:syntax`, `:filetype`

Reply:
@Jlll1 After some discussion on Matrix, I think we want to keep it simple and have `:trust` take no arguments and simply (1) calculate the hash of the current buffer contents (as they exist in memory, **not** on-disk), and (2) update the trust database with the calculated hash and file path.

* `:trust` returns an error for buffers with `buftype=nofile` or for which `nvim_buf_get_name()` returns an empty string
* In order to manually trust a file, a user can first `:edit` or `:view` it to open the buffer in Neovim, and then just use `:trust`
* To deny a file, use `:trust ++deny`

Also I think we should update the interface of `vim.secure.trust()` to accept a buffer number rather than a file path. There is always the possibility that the contents of a buffer do not match the contents of the file on disk, so by using buffer numbers we can ensure that the hash that is calculated is using the contents of the _buffer_, not the contents of the _file_ (a subtle but important difference).

Basically this means that `:trust` just runs `vim.secure.trust(<current buffer>, 'allow')`.

Reply:
Should we use `remove` instead of `forget` as per @justinmk comment? I like `remove` much better.

Reply:
Yes

Reply:
I also feel like splitting `trust` and `deny/forget` in the lua api is worth considering. From what I understand, `trust` is the only command where we care about the hash being in sync with what the user sees, so it makes sense for it to either accept `bufnr` or just take contents of current buffer. `deny/forget` on the other hand, do not need to receive `bufnr`, since `path` is more than enough. That means that the signature of the function would have to take in both `bufnr` and `path`, even though it won't use one of the args depending on the action. This could be confusing and lead to people trying to call `trust('allow', path)`, which then would either be super confusing and just take contents of current buffer, or just error out, since you're supplying `path` with `allow`. 
So my proposition would be to instead do `trust([optional]bufnr)`, `deny(path)` and `remove(path)`.

Reply:
Maybe we could switch the behavior based on type. Number for buffers, string for paths.

Reply:
Would the command be also optional then? E.g. `trust(0)` would imply `allow`, but `trust('foo', 'deny'|'remove')` would require the command? Either way we'd still have to implement errors on calls such as `trust(0, 'deny')`/
I slightly prefer separate functions, because then it's clear what's expected of you directly from the signature, while when having a single function, you need additional documentation. To @gpanders point of single `trust()` better conveying the message that this is the api to manage trust database, maybe `trust_allow`, `trust_remove` and `trust_deny`?

Reply:
I don't think the action/command should be optional, I'd keep everything explicit.

- `:trust` calls `trust('allow', 0)`
  Validate that it's not an unnamed buffer, and that there is a corresponding file on disk
- `:trust ++deny` calls `trust('deny', 0)`
  Buffer contents are ignored, just take the path of the current buffer
- `:trust ++deny file.txt` calls `trust('deny', 'file.txt')`
  The caveat with this is that we don't have any reliable way of expanding to absolute path if the file doesn't exist. Not a problem with `allow`, but it would be nice to just blacklist something even when there is nothing there yet. Maybe for now let's just do the buffers, and implement paths later on?
- Same for `++forget`/`++remove`

I'm personally okay with having multiple functions, that's why I thought a separate module would be better in the first place. I don't know about others.

Reply:
>I slightly prefer separate functions, because then it's clear what's expected of you directly from the signature, while when having a single function, you need additional documentation.

True, but then you also have to document each function, and anything that is common among the three functions has to be documented three separate times.

Using a single function also enables us to expand the interface in the future if we ever need to. Splitting it up into `allow`/`remove`/`deny` locks us into a less flexible position.

My vote is to make `trust()` take a single table argument. This table can then accept `bufnr`, `path`, and `action`, and the combination of inputs can be validated.

Examples:

```lua
-- OK, allows current buffer
vim.secure.trust({action = 'allow', bufnr = 0})

-- Error, 'allow' only works on buffers (not files)
vim.secure.trust({action = 'allow', path = ...})

-- OK, remove the file of the current buffer from the trust database
vim.secure.trust({action = 'remove', bufnr = 0})

-- OK, deny the given file
vim.secure.trust({action = 'deny', path = ...})

-- Error, bufnr and path are mutually exclusive
vim.secure.trust({action = 'remove', bufnr = 0, path = ...})
```

I'd like to hear what @justinmk thinks as well.

Reply:
For now I implemented it as suggested by @gpanders in 2fe2e19ee005d228b89bad38ac711df37adbbdc3